File: src/main/java/gherkin/TokenScanner.java
package gherkin;

import gherkin.ast.Location;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.Reader;
import java.io.StringReader;

public class TokenScanner {

    private BufferedReader reader;
    private int lineNumber;

    public TokenScanner(String source) {
        this(new StringReader(source));
    }

    public TokenScanner(Reader source) {
        this.reader = new BufferedReader(source);
    }

    public Token read() throws IOException {
        String line = reader.readLine();
        Location location = new Location(++lineNumber, 0);
        return line == null ? new Token(null, location) : new Token(new GherkinLine(line, lineNumber), location);
    }
}

File: dotnet/Gherkin/TokenScanner.cs
using System;
using System.IO;
using System.Linq;
using Gherkin.Ast;

namespace Gherkin
{
    public class TokenScanner : ITokenScanner
    {
        protected int lineNumber = 0;
        protected readonly TextReader reader;

        public TokenScanner(TextReader reader)
        {
            this.reader = reader;
        }

        public virtual Token Read()
        {
            var line = reader.ReadLine();
            var location = new Location(++lineNumber);
            return line == null ? new Token(null, location) : new Token(new GherkinLine(line, lineNumber), location);
        }
    }
}